### PR TITLE
net: gptp: fixes related to multiple ports management

### PR DESCRIPTION
**Pull request by nicogrx**
Hello, first let me introduce myself ;)
I am a software engineer working on automotive SoC that embed a TSN Ethernet switch. I am doing some tests using the Zephyr L2 gPTP stack. And well, I must thank all the contributors for this very nice job.

That is said, I met some issues especially whenever several gPTP ports are enabled:

1) Whenever my device under test behave as a transparent clock (so when SYNC/FUP messages must be forwarded from one gPTP slave port to another gPTP master port), I met an issue related to correction field of FUP message that was not properly updated (it was systematically set to 0 instead of taking link propagation delay & residence time into account). So  I did the following patch:

net: gptp: fix again computation of follow up correction field

2) I also noticed SYNC/FUP could be mistakenly generated by ClockMasterSyncSend state machine event if time aware system is not acting as the grand master clock (but is just a transparent clock). So this patch:

net: gptp: do not run clock master sync send SM if we are not the GM clock

3) Then, in case time aware system is the Grand Master clock, I saw that computation of scaled_rate_offset variable
    in gptp_md_follow_up_prepare() looks not correct. So this patch:

net: gptp: fix rate_ratio in gptp_mi_set_ps_sync_cmss()

4) Finally, if the device is acting as a Grand Master clock (while having several gPTP master ports as capable), I noticed that SYNC/FUP messages were only transmitted to one port after a couple of time. So I did this patch:

net: gptp: fix race condition on timestamp callback

While it seems to build on latest HEAD of main branch, I only tested it on a quite old revision of Zephyr kernel (that is compatible with my DUT).  Transparent-clock & GM clock test-cases with multiple ports of our DUT connected to certified tools looks ok at first sight.

But those patches may not be perfect for sure. So feel free to challenge of course!
Regards, Jean-Nicolas.

### CODE REVIEW

**Review by jukkar — COMMENTED** [review]
LGTM, only some cosmetic nits.

@chencaidy please take a look again if possible.

File: subsys/net/l2/ethernet/gptp/gptp_md.c
Comment (nicogrx) [reply subsys/net/l2/ethernet/gptp/gptp_md.c]: Hello Jukkar, I just did a push force of the gptp_fixes branch in my github. Hope is ok. Thanks for the review ;)

File: subsys/net/l2/ethernet/gptp/gptp_mi.c
Comment (nicogrx) [reply subsys/net/l2/ethernet/gptp/gptp_mi.c]: Hello Jukkar, I just did a push force of the gptp_fixes branch in my github. Hope is ok. Thanks for the review ;)

File: subsys/net/l2/ethernet/gptp/gptp_md.c
Comment (nicogrx) [reply subsys/net/l2/ethernet/gptp/gptp_md.c]: Sorry Jukkar, I had to push force again in order to remove "Change-ID" in commit messages... Sorry for that.

**Review by jukkar — APPROVED** [review]
Thanks, LGTM

**Comment by nicogrx** [comment]
Hi guys,

So I finally did a gPTP transparent clock test (relaying GM time in between gptp Slave port of DUT to a gPTP Master port), and monitoring time synchronization offset in between Master and Slave endpoint. Test is still running after 13 hours of testing with slave endpoint still synced to master endpoint.

gptp_transparent_clk_master_slave_offset_graph

Regards Jean-Nicolas.

**Comment by chencaidy** [comment]
> Hi guys,
> 
> So I finally did a gPTP transparent clock test (relaying GM time in between gptp Slave port of DUT to a gPTP Master port), and monitoring time synchronization offset in between Master and Slave endpoint. Test is still running after 13 hours of testing with slave endpoint still synced to master endpoint.
> 
> gptp_transparent_clk_master_slave_offset_graph
> 
> Regards Jean-Nicolas.

Thanks for the test, I'm in the middle of moving to another city right now, I'll rebuild the test platform to try it out when I'm done, Zypher does seem to be missing the PPS API at the moment, I think the MCU's 1588 timer (such as STMF4 and NXP RT10xx which tested) is also generally weak to programming PPS, this could be one of the reasons.

Here's my another PR: #61521 if you're interested in. After apply those patches, I use RT1052 as GM and STM32F407 as Endpoint, 802.11AS defaule profile(125ms sync interval), accuracy can be within 40ns.

**Comment by nicogrx** [comment]
> > Hi guys,
> > So I finally did a gPTP transparent clock test (relaying GM time in between gptp Slave port of DUT to a gPTP Master port), and monitoring time synchronization offset in between Master and Slave endpoint. Test is still running after 13 hours of testing with slave endpoint still synced to master endpoint.
> > gptp_transparent_clk_master_slave_offset_graph
> > Regards Jean-Nicolas.
> 
> Thanks for the test, I'm in the middle of moving to another city right now, I'll rebuild the test platform to try it out when I'm done, Zypher does seem to be missing the PPS API at the moment, I think the MCU's 1588 timer (such as STMF4 and NXP RT10xx which tested) is also generally weak to programming PPS, this could be one of the reasons.
> 
> Here's my another PR: #61521 if you're interested in. After apply those patches, I use RT1052 as GM and STM32F407 as Endpoint, 802.11AS defaule profile(125ms sync interval), accuracy can be within 40ns.

Hi Chen, No worry take your time. Thanks for the answer and for sharing #PR 61521.

Basically, without implementing any kind of PID algorithm in gptp_update_local_port_clock(), I tried to tune a bit that function. My PTP reference clock on our current (FPGA) setup is rated at a quite low frequency. So, I guess I cannot get something very accurate for the moment.

My DUT implements the full PTP clock. The only hack I changed in gptp_update_local_port_clock() to avoid getting high jitter, was to only call the fine correction method (=ptp_clock_rate_adjust()) and not the coarse one based on time increment (=ptp_clock_adjust()), this once time offset in between GM and local clock reached some minimum value. But this might not be optimal at all for sure. I will take a look to your patches. Thanks.

**Comment by nicogrx** [comment]
Hello, any updates about reviewing/merging this PR ? Thanks Jean-Nicolas.

**Comment by jukkar** [comment]
ping @chencaidy

**Comment by nicogrx** [comment]
Hello @chencaidy, did you find some time to cherry-pick this PR and check it works well at your side ?
Thanks.

File: subsys/net/l2/ethernet/gptp/gptp_mi.c
Comment (fgrandel) [inline subsys/net/l2/ethernet/gptp/gptp_mi.c:2031]: Thanks for providing the rationale of your change with pointers to the spec. This makes it so much easier to understand what this does.

File: subsys/net/l2/ethernet/gptp/gptp_mi.c
Comment (fgrandel) [inline subsys/net/l2/ethernet/gptp/gptp_mi.c:1]: A very small nitpick (not relevant for merging): If you like you could ensure that the main title of your commit message does not overflow for better readability (see e.g. 920558d71e2a4d519e829c9628fc6c1858fbb9f1)

File: subsys/net/l2/ethernet/gptp/gptp_mi.c
Comment (fgrandel) [inline subsys/net/l2/ethernet/gptp/gptp_mi.c:970]: Nice! Thank you for documenting these deficiencies. We might even open up a bug report for that. Looks like something that could be tagged as "[good first issue](https://github.com/zephyrproject-rtos/zephyr/labels/Good%20first%20issue)". If you like, you can do so and point to this comment. (Also not relevant for merging, just if you feel ambitious. ;-) )

**Review by ghost — APPROVED** [review]
First: So sorry that you had to wait so long for a second review. It would be great if we had more collaborators on PTP or even s/o able to work as a maintainer. If you want to grow into that position, let us know. I'd be happy to mentor you on your path (if that is even necessary at all - which it doesn't look like...).

In any case, many thanks for these valuable additions and the nice way in which you structured your commits. Really reviewer friendly.

**Pull request merged**
